### PR TITLE
Custom validation of server certificate for WebSockets

### PR DESCRIPTION
**Pull request by tintoy**
Hi.

This PR adds improved support for customisation of SSL-related options for WebSocket connections. It depends on `netcoreapp2.1` so these features are only available when consumers are targeting that. If they're targeting `netstandard`, the full .NET framework, or an earlier version of `netcoreapp` then the existing behaviour and API surface still apply.

**Note:** this PR (temporarily) adds some code (that I originally adapted for KubeClient) from .NET Core that sets up the WebSocket connection (once .NET Core 2.1 is RTM most of it can be removed).
.NET Core is MIT-licensed, which is more permissive than the Apache license so as far as I am aware this is acceptable (but if anyone else thinks differently, let me know). I have included an appropriate copyright notice and link to the .NET Core license text in the file (`CoreFX.cs`) that contains this code.

I also had to make a few small changes to the Travis build process; we need the .NET Core `2.1.300-preview1` SDK to build and run tests, so this is now installed as part of the build (Travis doesn't know how to install the preview SDK so we do it manually).

There is now a mock server that can handle WebSockets requests and the first test (exec-in-pod) that exercises WebSocket connection and communication.

Would probably be good to (at some stage) add some tests that wrap the WebSockets in a `StreamDemuxer`.

I've hooked up the test server's log to Xunit's test output (to make it easier to diagnose deadlocks etc) but haven't tried to integrate the `Microsoft.Rest` logging into that yet (figured that's probably a separate PR).

A fair bit of this code (especially the test and test-infrastructure) has been copied / adapted from [KubeClient](https://github.com/tintoy/dotnet-kube-client); hopefully it'll be useful to this project :)

Closes kubernetes-client/csharp#102

### CODE REVIEW

**Comment by tintoy** [comment]
CC: @qmfrederik @brendanburns

**Comment by tintoy** [comment]
It be a good idea to split out the `netcoreapp2.1` tests into a separate test project.

**Comment by tintoy** [comment]
Ugh, ok - looks like the test fails on Linux; I think for preview1, I need to explicitly opt into using the managed WebSocket implementation (by setting an environment variable).

**Comment by tintoy** [comment]
Also, looks like the CI script doesn't fail if tests fail. I'll see if I can fix that.

**Comment by tintoy** [comment]
Oh, it's actually 2 unrelated tests (`WatchTests.SuriveBadLine` and `WatchTests.TestWatchWithHandlers`) that are failing; interestingly, they only fail intermittently when I run them locally :)

**Comment by tintoy** [comment]
CC: @tg123 - is it possible that the build server is running the tests too slowly?

**Comment by tintoy** [comment]
They seem to be a little racy - they're passing now :)

**Comment by tintoy** [comment]
I've modified `MockKubeApiServer` to send logs to the test output (when configured to do so) so that the logs are only shown if a test fails. Individual tests can opt into this logging by injecting `ITestOutputHelper` into the test class constructor and passing it to the `MockKubeApiServer` constructor.

**Comment by tintoy** [comment]
I've tweaked some of the `WatcherTests` so they use a `CountDownEvent` to wait for incoming events.
They pass now, but one of the `ByteBufferTests` is failing instead (seems like the same issue). I'll see what I can do there as well.

**Comment by tintoy** [comment]
Ok, there seem to be a couple of racy tests left, but I think I'll leave it for now; there are already more changes than I'm comfortable making in a single PR :wink:

**Comment by brendandburns** [comment]
Thanks! Will review tomorrow, I kicked the CI/CD again...

**Comment by tintoy** [comment]
Sorry in advance for the amount of code in this PR - once I broke the build, I just got stubborn :grin:

**Comment by tintoy** [comment]
@brendanburns - FYI, it looks like the change for custom validation of server certificate may wind up making it into 2.1 after all: https://github.com/dotnet/corefx/issues/12038#issuecomment-374303655

If this is still merged (the separate code for `netstandard` vs `netcoreapp2.1` is still required), I guess it can be refactored when 2.1 goes RTM?

File: ci.sh [hunk 1]
```diff
@@ -4,9 +4,12 @@
...
-find . -name *.csproj -exec dotnet build {} \;
+dotnet restore
+dotnet build --no-restore
 
 # Execute Unit tests
 cd tests
-dotnet restore
-dotnet test
+dotnet test --no-restore --no-build
+if [[ $? != 0 ]]; then
```
Comment (tg123) [inline ci.sh]: already set -e

Comment (tintoy) [reply ci.sh]: Yeah, I was wondering about that - it's supposed to stop on error (if you have set -e) right? But if I take that out then the build doesn't fail if tests fail :-/

Comment (tintoy) [reply ci.sh]: I'm no bash expert so it's entirely possible I've got it wrong though.

Comment (tintoy) [reply ci.sh]: I can't remember where, but I remember reading something a while back that said "set -e" is unreliable in some cases and you're better off using "trap" but again, I could be wrong :)

File: src/WebSocketBuilder.NetCoreApp2.1.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            Options.ClientCertificates.Add(certificate);
+
+            return this;
+        }
+
+        public WebSocketBuilder ExpectServerCertificate(X509Certificate2 serverCertificate)
+        {
+            Options.ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) =>
+            {
+                if (sslPolicyErrors != SslPolicyErrors.RemoteCertificateChainErrors)
```
Comment (brendandburns) [inline src/WebSocketBuilder.NetCoreApp2.1.cs]: nit: curly braces around all `if` ...

**Comment by brendandburns** [comment]
LGTM.
This looks good enough to me and this PR has grown sufficiently large that I'd like to get it merged and then iterate...

**Comment by tintoy** [comment]
Thanks! I'll open a new PR to address your feedback :)

File: src/WebSocketBuilder.NetCoreApp2.1.cs (hunk 2 above)
Comment (tintoy) [reply src/WebSocketBuilder.NetCoreApp2.1.cs]: Gotcha, will make sure I do from now on :)

**Pull request merged**